### PR TITLE
`generate_mil_features` for extracting last layer activations in TransMIL and AttentionMIL models

### PR DESCRIPTION
**Pull request by CostanzaSiniscalchi**
# Description

Added new function `generate_mil_features` that returns a MILActivations object generated from extracting activation weights from the last layer of a TransMIL or AttentionMIL model. This feature could allow the user to use activation weights as input to a fusion model.
Required dependencies: None
Args:
        weights (str): Path to model weights to load.
        dataset (sf.Dataset): Dataset to evaluation.
        outcomes (str, list(str)): Outcomes.
        bags (str, list(str)): Path to bags, or list of bag file paths.
            Each bag should contain PyTorch array of features from all tiles in
            a slide, with the shape ``(n_tiles, n_features)``.
        config (:class:`slideflow.mil.TrainerConfigFastAI` or :class:`slideflow.mil.TrainerConfigCLAM`):
            Configuration for building model. If ``weights`` is a path to a
            model directory, will attempt to read ``mil_params.json`` from this
            location and load saved configuration. Defaults to None.

As part of this addition, the function `predict_mil_activations` has been created in slideflow.mil.eval, which is a copy of `predict_mil`, but also calls new method `get_last_layer_activations` from the given model class.
`get_last_layer_activations` in TransMIL and AttentionMIL classes is a copy of the `forward` method, but simply returns the layer weights before they are converted to scores.

## Type of change
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

- Trained a TransMIL model using Ctranspath features generated from a 52-person dataset (PEOPLE at 20x magnification). Used the path to the model weights as input to `generate_mil_features` along with the path to the features as bags, the same dataset the model was trained on, and the same configuration as the model.
- Trained a AttentionMIL model using Ctranspath features generated from a 52-person dataset (PEOPLE). Used the path to the model weights as input to `generate_mil_features` along with the path to the features as bags, the same dataset the model was trained on, and the same configuration as the model.

### CODE REVIEW

File: slideflow/mil/models/att_mil.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ def __init__(
             encoder:  A network transforming bag instances into feature vectors.
         """
         super().__init__()
+        self.n_feats= n_feats
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:28]: Make sure these two line are used and delete them

File: slideflow/mil/models/att_mil.py [hunk 2]
```diff
@@ -25,6 +25,8 @@ def __init__(
...
+        self.n_classes= n_out
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:29]: line 2

File: slideflow/project.py [hunk 3]
```diff
@@ -38,11 +38,12 @@
     get_first_nested_directory, get_matching_directory, BreastER, ThyroidBRS,
     LungAdenoSquam
 )
+from .mil._params import _TrainerConfig
 
 if TYPE_CHECKING:
     from slideflow.model import DatasetFeatures, Trainer
     from slideflow.slide import SlideReport
-    from slideflow import simclr, mil
+    from slideflow import simclr, swav, mil
```
Comment (matte-esse) [inline slideflow/project.py:46]: remove swav from here, I don't know how it ended up here

File: slideflow/project.py [hunk 4]
```diff
@@ -3677,6 +3688,119 @@ def train_simclr(
         )
         simclr.run_simclr(simclr_args, builder, model_dir=outdir, **kwargs)
 
+    def train_swav_old(
```
Comment (matte-esse) [inline slideflow/project.py:3691]: remove the function

File: test.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import slideflow as sf
```
Comment (matte-esse) [inline test.py:1]: remove this file

File: slideflow/mil/eval.py [hunk 6]
```diff
@@ -10,10 +10,12 @@
 from slideflow import Dataset, log, errors
 from slideflow.util import path_to_name
 from slideflow.stats.metrics import ClassifierMetrics
+from .features import MILActivations
 from ._params import (
     _TrainerConfig, ModelConfigCLAM, TrainerConfigCLAM
 )
 
+
```
Comment (matte-esse) [inline slideflow/mil/eval.py:18]: remove line

File: slideflow/mil/features.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+from collections import defaultdict
+from typing import Union, List, Optional, Callable, Tuple, Any
+
+import numpy as np
+import pandas as pd
+from ._params import (
+    _TrainerConfig, ModelConfigCLAM, TrainerConfigCLAM
+)
+
+class MILActivations():
```
Comment (matte-esse) [inline slideflow/mil/features.py:13]: add `from_df`,

File: slideflow/mil/features.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
```
Comment (matte-esse) [inline slideflow/mil/features.py:14]: (add `sf.SlideMap.from_mil_activations`)
here add `map_activations` which calls calls `sf.SlideMap.from_mil_activations`

File: slideflow/mil/features.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self, model: Callable, config: _TrainerConfig, hlw: list, slides: list, annotations: pd.DataFrame):
```
Comment (matte-esse) [inline slideflow/mil/features.py:15]: add documentation in all functions

File: slideflow/mil/features.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        self.model= model
```
Comment (matte-esse) [inline slideflow/mil/features.py:16]: run `shift + alt + F` on files before you push them

File: slideflow/mil/models/transmil.py [hunk 11]
```diff
@@ -8,6 +8,8 @@
 class TransMIL(nn.Module):
     def __init__(self, n_feats: int, n_out: int,):
         super(TransMIL, self).__init__()
+        self.n_feats= n_feats
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:11]: remove this and next line

File: slideflow/mil/eval.py [hunk 12]
```diff
@@ -425,3 +532,73 @@ def _predict_mil(
             y_pred.append(torch.nn.functional.softmax(model_out, dim=1).cpu().numpy())
     yp = np.concatenate(y_pred, axis=0)
     return yp, y_att
+
+
+def _predict_mil_activations(
```
Comment (matte-esse) [inline slideflow/mil/eval.py:537]: `_get_mil_activations`

File: slideflow/mil/eval.py [hunk 13]
```diff
@@ -425,3 +532,73 @@ def _predict_mil(
...
+        loaded = torch.load(bag).to(device)
+        loaded = torch.unsqueeze(loaded, dim=0)
+        with torch.no_grad():
+            if use_lens:
+                lens = torch.from_numpy(np.array([loaded.shape[1]])).to(device)
+                model_args = (loaded, lens)
+            else:
+                model_args = (loaded,)
+            model_out = model(*model_args)
+            h = model.get_last_layer_activations(*model_args)#added
```
Comment (matte-esse) [inline slideflow/mil/eval.py:582]: this function should be in all models, from the models that don't have it implemented yet it should raise an exception `function not implemented` or similar

File: slideflow/project.py [hunk 14]
```diff
@@ -38,6 +38,7 @@
     get_first_nested_directory, get_matching_directory, BreastER, ThyroidBRS,
     LungAdenoSquam
 )
+from .mil._params import _TrainerConfig
```
Comment (matte-esse) [inline slideflow/project.py:41]: let's not import this and make the line that uses it consist with train_mil which simply has a string

File: slideflow/project.py [hunk 15]
```diff
@@ -3677,6 +3704,64 @@ def train_simclr(
         )
         simclr.run_simclr(simclr_args, builder, model_dir=outdir, **kwargs)
 
+    def train_swav(
```
Comment (matte-esse) [inline slideflow/project.py:3707]: remove this function

File: slideflow/mil/models/transmil.py [hunk 16]
```diff
@@ -16,7 +16,7 @@ def __init__(self, n_feats: int, n_out: int,):
         self.layer2 = TransLayer(dim=512)
         self.norm = nn.LayerNorm(512)
         self._fc2 = nn.Linear(512, self.n_classes)
-
+
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:19]: remove 2 tabs 😂

File: slideflow/mil/models/transmil.py [hunk 17]
```diff
@@ -74,6 +74,36 @@ def forward(self, h):
...
+        h = self.pos_layer(h, _H, _W) #[B, N, 512]
+
+        #---->Translayer x2
+        h = self.layer2(h) #[B, N, 512]
+
+        #---->cls_token
+        h = self.norm(h)[:,0]
+        
+        return h
+
```
Comment (matte-esse) [inline slideflow/mil/models/transmil.py:104]: change it from 3 to 1 spacing line

File: slideflow/mil/features.py [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+        annotations: Union[str, "pd.core.frame.DataFrame"],
+        use_lens: bool,
+        config: Optional[_TrainerConfig] = None,
+        dataset: Optional["sf.Dataset"] = None,
+        outcomes: Optional[Union[str, List[str]]] = None,
+        attention_pooling: Optional[str] = 'avg',
+        device: Optional[Any] = None
+    ):
+        """Loads in model from Callable or path to model weights and config.
+        Saves activations from hidden layer weights, predictions, and attention weight,
```
Comment (matte-esse) [inline slideflow/mil/features.py:37]: 80th column line wrapping throughout the file

File: slideflow/mil/eval.py [hunk 19]
```diff
@@ -249,6 +252,107 @@ def predict_from_model(
...
+    log.info(f"Loading model weights from [green]{weights}[/]")
+    model.load_state_dict(torch.load(weights))
+
+    # Prepare device.
+    if hasattr(model, 'relocate'):
+        model.relocate()  # type: ignore
+    model.eval()
+
+    annotations = dataset.annotations
+    activations = MILFeatures(
```
Comment (matte-esse) [inline slideflow/mil/eval.py:351]: change var name to features

File: slideflow/mil/features.py [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+            obj.attentions = {
+                s: np.stack(df.loc[df.slide == s].attentions.values[0])
+                for s in obj.slides
+            }
+        if annotations:
+            obj.annotations = obj._get_annotations(annotations)
+
+        return obj
+
+    def to_df(self) -> pd.core.frame.DataFrame:
```
Comment (matte-esse) [inline slideflow/mil/features.py:333]: def to_df(self, activations=True, predictions=True)

If these are set to fals the dataframe should not have them

**Comment by matte-esse** [comment]
Very minor comments. Overall this is very good work @CostanzaSiniscalchi

File: slideflow/mil/features.py (hunk 20 above)
Comment (CostanzaSiniscalchi) [reply slideflow/mil/features.py:333]: Made changes to this function which will make it cleaner and adds CLAM model functionality

Comment (matte-esse) [reply slideflow/mil/features.py:384]: it still doesn't have `activations=True, predictions=True`

File: slideflow/project.py [hunk 21]
```diff
@@ -2264,6 +2264,33 @@ def generate_heatmaps(
...
+        outcomes: Union[str, List[str]],
+        bags: Union[str, np.ndarray, List[str]]
+    ):
+        """Generate activations weights from the last layer of an MIL model.
+
+        Returns MILFeatures object.
+
+        Args:
+            weights (str): Path to model weights to load.
+            dataset (sf.Dataset): Dataset to evaluation.
```
Comment (matte-esse) [inline slideflow/project.py:2280]: change this to

:class:`slideflow.Dataset`

please check this through your new code

File: slideflow/mil/features.py [hunk 22]
```diff
@@ -0,0 +1,458 @@
...
+        annotations: Union[str, "pd.core.frame.DataFrame"],
+        config: Optional[_TrainerConfig] = None,
+        dataset: Optional["sf.Dataset"] = None,
+        outcomes: Optional[Union[str, List[str]]] = None,
+        attention_pooling: Optional[str] = 'avg',
+        device: Optional[Any] = None
+    ):
+        """Loads in model from Callable or path to model weights and config.
+        Saves activations from hidden layer weights, predictions, and attention weight, 
+        storing to internal parameters ``self.activations``, ``self.predictions``, ``self.attentions`` dictionaries
```
Comment (matte-esse) [inline slideflow/mil/features.py:38]: specify that this will be the last hiddne layer

File: slideflow/mil/features.py [hunk 23]
```diff
@@ -0,0 +1,458 @@
...
+        outcomes: Union[str, List[str]],
+        bags: Union[str, np.ndarray, List[str]]
+    ):
+        """Generate model from model path and config.
+
+        Returns callable model object.
+
+        Args:
+            weights (str): Path to model weights to load.
+            dataset (sf.Dataset): Dataset to evaluation.
```
Comment (matte-esse) [inline slideflow/mil/features.py:94]: :class etc

File: slideflow/mil/eval.py [hunk 24]
```diff
@@ -10,12 +10,14 @@
 from slideflow import Dataset, log, errors
 from slideflow.util import path_to_name
 from slideflow.stats.metrics import ClassifierMetrics
+from .features import MILFeatures
 from ._params import (
     _TrainerConfig, ModelConfigCLAM, TrainerConfigCLAM
 )
 
 # -----------------------------------------------------------------------------
```
Comment (matte-esse) [inline slideflow/mil/eval.py:19]: remove line

File: slideflow/project.py [hunk 25]
```diff
@@ -2270,6 +2273,34 @@ def generate_heatmaps(
...
+        Returns MILFeatures object.
+
+        Args:
+            weights (str): Path to model weights to load.
+            dataset (:class:`slideflow.Dataset`): Dataset.
+            outcomes (str, list(str)): Outcomes.
+            bags (str, list(str)): Path to bags, or list of bag file paths.
+                Each bag should contain PyTorch array of features from all tiles
+                in a slide, with the shape ``(n_tiles, n_features)``.
+            config (:class:`slideflow.mil.TrainerConfigFastAI` or
```
Comment (matte-esse) [inline slideflow/project.py:2294]: order doesn't match the function definitino

File: slideflow/mil/features.py [hunk 26]
```diff
@@ -0,0 +1,478 @@
...
+
+        Args:
+            model (str): Path to model from which to calculate activations.
+            bags (str, list): Path or list of feature bags,
+            slides (list): List of slides, 
+            annotations: Union[str, "pd.core.frame.DataFrame"],
+            use_lens (bool): Spec used for generating model args in 
+            _get_mil_activations,
+        Keyword Args:
+            config (:class: `TrainerConfig`]: Trainer for MIL model,
```
Comment (matte-esse) [inline slideflow/mil/features.py:53]: remove space between colon and backtick

File: slideflow/mil/features.py [hunk 27]
```diff
@@ -0,0 +1,478 @@
...
+        outcomes: Union[str, List[str]],
+        bags: Union[str, np.ndarray, List[str]]
+    ):
+        """Generate model from model path and config.
+
+        Returns callable model object.
+
+        Args:
+            weights (str): Path to model weights to load.
+            dataset (sf.Dataset): Dataset to evaluation.
```
Comment (matte-esse) [inline slideflow/mil/features.py:101]: change sf.Dataset

File: slideflow/mil/features.py [hunk 28]
```diff
@@ -0,0 +1,478 @@
...
+            outcomes (str, list(str)): Outcomes.
+            bags (str, list(str)): fPath to bags, or list of bag file paths.
```
Comment (matte-esse) [inline slideflow/mil/features.py:103]: typo fPath

File: slideflow/mil/features.py [hunk 29]
```diff
@@ -0,0 +1,478 @@
...
+        config: _TrainerConfig,
+        dataset: "sf.Dataset",
+        outcomes: Union[str, List[str]],
+        bags: Union[str, np.ndarray, List[str]]
+    ):
+        """Generate model from model path and config.
+
+        Returns callable model object.
+
+        Args:
```
Comment (matte-esse) [inline slideflow/mil/features.py:99]: order of Args in documentations doesn't match function def order

File: slideflow/mil/features.py [hunk 30]
```diff
@@ -0,0 +1,478 @@
...
+            weights (str): Path to model weights to load.
+            dataset (sf.Dataset): Dataset to evaluation.
+            outcomes (str, list(str)): Outcomes.
+            bags (str, list(str)): fPath to bags, or list of bag file paths.
+                Each bag should contain PyTorch array of features from all tiles
+                in a slide, with the shape ``(n_tiles, n_features)``.
+            config (:class:`slideflow.mil.TrainerConfigFastAI` or 
+            :class:`slideflow.mil.TrainerConfigCLAM`):
```
Comment (matte-esse) [inline slideflow/mil/features.py:107]: tab

File: slideflow/mil/features.py [hunk 31]
```diff
@@ -0,0 +1,478 @@
...
+        device: Optional[Any]
+    ):
+        """Loads in model from Callable and calculates activations, predictions,
+        and attention weights.
+
+        Args:
+            model (Callable): Model from which to calculate activations.
+            bags (list): List of feature bags,
+            use_lens (bool): Spec used for generating model args in 
+            _get_mil_activations, generate activations.
```
Comment (matte-esse) [inline slideflow/mil/features.py:208]: tab

File: slideflow/mil/features.py [hunk 32]
```diff
@@ -0,0 +1,478 @@
...
+        model: Callable,
+        bags: Union[np.ndarray, List[str]],
+        attention_pooling: str,
+        use_lens: bool,
+        device: Optional[Any]
+    ):
+        """Loads in model from Callable and calculates activations, predictions,
+        and attention weights.
+
+        Args:
```
Comment (matte-esse) [inline slideflow/mil/features.py:204]: order

File: slideflow/mil/features.py [hunk 33]
```diff
@@ -0,0 +1,478 @@
...
+        num_features, acts = self._get_activations(hs)
+        atts = self._get_attentions(y_att)
+        preds = self._get_predictions(yp)
+
+        return num_features, preds, atts, acts
+
+    def _get_activations(self, hlw):
+        """Formats list of activation weights into dictionary of lists, 
+        matched by slide.
+        Returns the dictionary."""
```
Comment (matte-esse) [inline slideflow/mil/features.py:289]: remove `Returns the dictionary."""`. In general you can be a bit less specific - and sometime even omit documentation - in private methods

File: slideflow/mil/eval.py [hunk 34]
```diff
@@ -156,80 +189,6 @@ def eval_mil(
...
```
Comment (matte-esse) [inline slideflow/mil/eval.py:191]: delete method, bring it back

File: slideflow/mil/eval.py [hunk 35]
```diff
@@ -363,92 +426,6 @@ def generate_attention_heatmaps(
...
```
Comment (matte-esse) [inline slideflow/mil/eval.py:428]: deleted method, bring it back

File: slideflow/project.py [hunk 36]
```diff
@@ -28,6 +28,7 @@
 from typing import (TYPE_CHECKING, Any, Callable, Dict, List, Optional, Tuple,
                     Union)
 
+
```
Comment (matte-esse) [inline slideflow/project.py:31]: delete line

File: slideflow/project.py [hunk 37]
```diff
@@ -1451,7 +1452,8 @@ def evaluate_mil(
         Logs classifier metrics (AUROC and AP) to the console.
 
         Args:
-            config (:class:`slideflow.mil.TrainerConfigFastAI` or :class:`slideflow.mil.TrainerConfigCLAM`):
+            config (:class:`slideflow.mil.TrainerConfigFastAI` or 
+            :class:`slideflow.mil.TrainerConfigCLAM`):
```
Comment (matte-esse) [inline slideflow/project.py:1456]: this has got nothing to do with this PR but could you please fix the args in the function train_dataset, val_dataset to dataset

File: slideflow/mil/features.py [hunk 38]
```diff
@@ -0,0 +1,475 @@
...
+        attention weight, storing to internal parameters ``self.activations``,
+         ``self.predictions``, ``self.attentions`` dictionaries 
+        mapping slides to arrays of weights.
+        Converts annotations to DataFrame if necessary.
+
+        Args:
+            model (str): Path to model from which to calculate activations.
+            bags (str, list): Path or list of feature bags,
+            slides (list): List of slides, 
+            annotations: Union[str, "pd.core.frame.DataFrame"],
```
Comment (matte-esse) [inline slideflow/mil/features.py:49]: order or arguments does not match function def (check all functions)

File: slideflow/mil/features.py [hunk 39]
```diff
@@ -0,0 +1,475 @@
...
+        else:
+            raise RuntimeError(
+                'Model path detected without config, dataset, bags, or \
+                    outcomes')
+        if self.model:
+            self.num_features, self.predictions, self.attentions, \
+                self.activations = self._get_mil_activations(
+                    self.model, bags, attention_pooling, use_lens, device)
+
+    def generate_model(
```
Comment (matte-esse) [inline slideflow/mil/features.py:87]: this should be a private method

File: slideflow/mil/eval.py [hunk 40]
```diff
@@ -294,6 +303,108 @@ def predict_from_model(
...
+    dataset: "sf.Dataset",
+    outcomes: Union[str, List[str]],
+    bags: Union[str, np.ndarray, List[str]]
+):
+    """Generate activations weights from the last layer of an MIL model.
+
+    Returns MILFeatures object.
+
+    Args:
+        weights (str): Path to model weights to load.
```
Comment (matte-esse) [inline slideflow/mil/eval.py:318]: order

File: slideflow/mil/models/clam.py [hunk 41]
```diff
@@ -298,6 +303,68 @@ def __init__(
         # Initialize weights.
         initialize_weights(self)
 
+    def get_last_layer_activations(
```
Comment (matte-esse) [inline slideflow/mil/models/clam.py:306]: move this to the upperclass

File: slideflow/mil/models/att_mil.py [hunk 42]
```diff
@@ -25,6 +25,7 @@ def __init__(
             encoder:  A network transforming bag instances into feature vectors.
         """
         super().__init__()
+
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:28]: delete line

File: slideflow/mil/models/att_mil.py [hunk 43]
```diff
@@ -38,12 +39,24 @@ def forward(self, bags, lens):
         embeddings = self.encoder(bags)
 
         masked_attention_scores = self._masked_attention_scores(embeddings, lens)
+
```
Comment (matte-esse) [inline slideflow/mil/models/att_mil.py:42]: delete line

File: slideflow/project.py (hunk 37 above)
Comment (CostanzaSiniscalchi) [reply slideflow/project.py:1456]: sure. I also fixed the order of the args to match the way they are in the definition

**Comment by matte-esse** [comment]
please correct the body of the PR so that it matches the changes that were done (mostly the functions names)

**Pull request merged**